[#69] Allow `flush` to accepts an amout of proposals, add `dropProposals` entrypoint

rinn7e opened the pull request:
## Description
Problem: When calling `flush`, `decisionLambda` or `rejectedValue` lambda may fail due to unexpected error. We should have a way to drop some proposals and a way to not flush proposals all at one.

Solution: 
- Refactor `sProposalKeyListSortByDate` to be a map with `(Timestamp, ProposalKey proposalMetadata)` as the key, and `()` as the value (Since value here is not important).
- Add `dropProposals` entrypoint
- Update `flush` entrypoint to accept a (Maybe Natural)
- Add tests for these new features

<!--
Describes the nature of your changes. If they are substantial, you should
further subdivide this into a section describing the problem you are solving and
another describing your solution.
-->

## Related issue(s)

<!--
- Short description of how the PR relates to the issue, including an issue link.
For example
- Fixed #100500 by adding lenses to exported items

Write 'None' if there are no related issues (which is discouraged).
-->

Resolves #69

## :white_check_mark: Checklist for your Pull Request

<!--
Ideally a PR has all of the checkmarks set.

If something in this list is irrelevant to your PR, you should still set this
checkmark indicating that you are sure it is dealt with (be that by irrelevance).

If you don't set a checkmark (e. g. don't add a test for new functionality),
you must be able to justify that.
-->

#### Related changes (conditional)

- Tests
  - [x] If I added new functionality, I added tests covering it.
  - [x] If I fixed a bug, I added a regression test to prevent the bug from
        silently reappearing again.

[//]: # (Add more docs here if you have them in the repository)
- Documentation
  - [x] I checked whether I should update the docs and did so if necessary:
    - [README](../tree/master/README.md)
    - Haddock

#### Stylistic guide (mandatory)

- [x] My commits comply with [the following policy](https://www.notion.so/serokell/Commit-and-PR-policy-4cf98e1a910a415d86b5f2491d9af1af).
- [x] My code complies with the [style guide](../tree/master/docs/code-style.md).

rinn7e commented on src/Lorentz/Contracts/BaseDAO/Types.hs:
```diff
@@ -217,7 +217,7 @@ data Storage (contractExtra :: Kind.Type) (proposalMetadata :: Kind.Type) = Stor
   , sExtra                     :: contractExtra
   , sProposals                 :: BigMap (ProposalKey proposalMetadata) (Proposal proposalMetadata)
-  , sProposalKeyListSortByDate :: [ProposalKey proposalMetadata] -- Newest first
+  , sProposalKeyListSortByDate :: Map Timestamp (ProposalKey proposalMetadata) -- Oldest first
```
This MR is still WIP, currently allow `flush` to accepts an amount of proposals is done. I will work on adding `dropProposals` entrypoint and their tests next.

One thing that I need advice on is that, currently, I have to update `sProposalKeyListSortByDate` from a list to map of proposal created date and its key.

This is due to 2 main reasons:
- There is currently a bug in `handleProposalIsOver`: https://github.com/tqtezos/baseDAO/blob/master/src/Lorentz/Contracts/BaseDAO/Proposal.hs#L656-L663
  - I implemented this naively, this end up reversing the list. 
  Since we process all the proposals all at the same time, it was hard to detect this.

- Using a list is hard to modify. Since we can handle certain amount of a proposals now, using `Map` is much easier to update/remove the proposal key.

Hope to get your views on this @gromakovsky @Martoon-00 .

Martoon-00 replied:
I remember that in another project we also sometimes had to use maps for structures that were initially list-like - because of some Michelson limitations. So I think it's a good change if it simplifies things.

However, note that keeping `Map Timestamp a` may be tricky since `NOW` eventually returns the time of operation inclusion into a block, thus probability of collision is high. You will probably have to use `Map Timestamp [a]`. And if so, I wonder whether will we get any benefits from using Map. 

Another solution is to use numeric indices as keys and remember the indices of the first and last active proposals or something like that. And add an index in that map into `Proposal` datatype. This is gonna be even less pleasant than working with `Map Timestamp` though, but the most correct and flexible.

Could you elaborate on the second point: "Using a list is hard to modify"? AFAIU the first point is caused by the lack of tests on storage (#31) and is quite fixable. Regarding the second point, probably the following dareful plain would be possible to implement:
* Keep the list in "oldest goes first" order;
* Now we don't even need to limit the number of active proposals, because when it is high, `propose` will be the most expensive operation and thus the number of proposals will become naturally limited.
* `ensureProposalIsUnique` can probably be implemented via seeking by proposal hash in proposals map;
* `addProposal` will add to the end of the list;
* `flush` will do a series of `IF_CONS` to consume the list in direct order.

This plan probably has its own issues, but if everything is agreed it seems the simplest to implement to me.

What do you think?

rinn7e replied:
> probability of collision is high

I didn't think of this, and this could be a big issue.

> Using a list is hard to modify
This is pretty much due to having to copy to a new list when we add/remove an element form a list. Since it is actually linear, I guess it is not a performance issue. 

I'm good with your points, and I will make these changes instead of using `Map`.

Martoon-00 replied:
> This is pretty much due to having to copy to a new list when we add/remove an element form a list. It is simply a bit tricky, and I was worried about the performance. But in our case, it seems to be linear anyway, I guess that wouldn't be a performance problem.

Yeah, right. Well, if we would use `BigMap Index Proposal`, then we could really have `O(log N)` cost of `propose` and `O(n)` cost for `flush(n)`, though this would be trickier to support.

rinn7e replied:
The issue with `BigMap` is that I don't think we can loop through it using `ITER`.

Related to the collision, I'm thinking of using `Map (Timestamp, ProposalKey) ()`, this way we can guarantee that it has no collision, it should also be sorted with timestamp as well, since it seems in michelson, it compare the left of a pair first.

With this in mind, do you think using `Map` could be better? If not, I'll still go with List.

Martoon-00 replied:
> The issue with BigMap is that I don't think we can loop through it using ITER.

Yep, but if we store indices of the first and last active proposals, then in all cases we can iterate by indices, in theory.

> Map (Timestamp, ProposalKey) ()

And then `ITER` over it to get proposals in appropriate order? Right this should work now. I though that there is no way to get all elements with timestamp less than a given value from such a map, but apparently `ITER` suits that.

Okay, I think it's reasonable to use map, should be safer at least, and modifying it is indeed much simpler.

rinn7e commented:
@gromakovsky @Martoon-00  This MR is ready for review now. 

A few things to note:
1. I have updated `sProposalKeyListSortByDate` to be `Map (Timestamp, ProposalKey proposalMetadata) ()`. 
	- This avoid collision, provide faster add/delete.
	- It is sorted by `Timestamp` from oldest to newest first, then ordered by the proposal key `bytestring` 
2. I have added `pIsDeleted` for proposal, to detect whether or not a proposal has been deleted. User that vote on a deleted proposal will get 'Proposal not exist' error, and flush now will mark all ended proposals as deleted.

3. This is not really necessary, but `dropProposals` should not accept 0 as param, so I added a check which will throw 'bad parameter' error. This applies to when pass `Just 0` to flush entrypoint as well.

> Note that as part of this issue we should also specify the order in which proposals are executed. `flush` can execute multiple proposals and the order may matter. The tricky thing here is that we can't just say "the earliest proposal is executed first" because 2 `propose` calls can be included in the same block and have the same `NOW` timestamp.

Related to your comment, with the new update, `flush` will execute base on the timestamp first, then the bytestring generated by the ProposeParams. I'm not sure how to put this into a proper word, and it is hard to predict the position of this bytestring as well.

Martoon-00 commented on src/Lorentz/Contracts/BaseDAO/Proposal.hs:
```diff
@@ -604,63 +581,153 @@ unfreezeVoterToken = do
+    , fieldCtor $ getField #pProposerFrozenToken
+    , fieldCtor $ getField #pVoters
+    , fieldCtor $ push True
+    )
```
Can't we use just `setField` here? Invoking getters so many times is indeed expensive, maybe avoiding this would save us a lot of operation size.

rinn7e replied:
I did it that where since I actually don't know `setField` exist. I have updated this to use `setField` now. 

Tho, he operation size is reduced to 14880, the test still failed.

Martoon-00 commented on src/Lorentz/Contracts/BaseDAO/Proposal.hs:
```diff
@@ -727,3 +735,93 @@ flush config = do
+    failCustom_ #vOTING_PERIOD_OVER -- TODO: forbidden entrypoint parameter
+  else
+    nop
+
```
Yep, these checks look like a good improvement to the contract semantics. 

Though a bit sad that for such things we pay with operation size which is a problem now. I suppose we could also offload such checks to the middleware via mentioning in the description something like "passing 0 is semantically no-op but is still expensive, please don't call us with 0".

Martoon-00 commented:
[This place](https://github.com/tqtezos/baseDAO/pull/71/commits/e3b94c63be8e2df0ee90739e5858b81bda711842#diff-4f7aa07adda34fc46cdfc608faad70f7e9c4eb7a1231d1eaafb34eb564770c09R285) looks like one more candidate for optimization, probably we could use just 3 setters there.

Martoon-00 commented:
Could you explain the difference between `pIsDeleted` flag and removing a proposal from the map? Like, previously we could forget about terminated proposals, but why `pIsDeleted` flag wasn't necessary where and got useful only now?

Martoon-00 commented on src/Lorentz/Contracts/BaseDAO/Proposal.hs:
```diff
@@ -604,63 +577,104 @@ unfreezeVoterToken = do
-    dip $ drop @(ProposalKey pm)
-
+      deleteProposalKeyFromMap
+      dip drop
```
Maybe we could also optimize here?: these two `else` branches seem to be identical but pretty big. Maybe we could construct a complex condition using `and`/`or` and have only one `if ... then ... else`? That should save a really lot of space.

Though a sad truth is that Lorentz does not have operators like `&&` and adding them there will take too long (no one severely used this functionality until baseDAO, I added [one](https://gitlab.com/morley-framework/morley/-/issues/505) and [two](https://gitlab.com/morley-framework/morley/-/issues/506) now), but probably we could add a helper that handles our specific condition.

rinn7e commented:
> Could you explain the difference between `pIsDeleted` flag and removing a proposal from the map? Like, previously we could forget about terminated proposals, but why `pIsDeleted` flag wasn't necessary where and got useful only now?

Actually, we probably don't need `pIsDeleted`. I make a mistake by not thinking about `dropProposals` should only deleted ended proposals that have not been flush. I'll make the fix and come back to this.

rinn7e commented on src/Lorentz/Contracts/BaseDAO/Proposal.hs:
```diff
@@ -727,3 +735,93 @@ flush config = do
+      dip $ dip drop
+      fromNamed #counter; push (1 :: Natural); add; toNamed #counter
+
+  drop; drop
```
Yes, currently this is a batch delete. Before it was affecting ongoing proposals, but now this should only delete finished proposals. 

Since the goal was simply to remove stuck proposals that fail due to decisionLambda, we probably will use this entrypoint to drop the first one, eg. `dropProposals 1`.

I agree that passing `ProposalKey` is better, so if you think we should really do it, I'll make the change. 

cc @gromakovsky

Martoon-00 replied:
Probably my inial proposal about dropping the head proposal was not good. Though let's really ask @gromakovsky.

gromakovsky replied:
What if we pass only one `ProposalKey`, compare it with the head one and remove the head one if they match (failing otherwise)? It should simplify the contract and make it smaller without significantly complicating its usage.

rinn7e replied:
> What if we pass only one `ProposalKey`, compare it with the head one and remove the head one if they match (failing otherwise)? It should simplify the contract and make it smaller without significantly complicating its usage.

I'm fine with this as well. If we're good, I'll make the changes.

rinn7e replied:
I'm thinking that in some scenario where we know that the proposal is going to be accepted, but containing fail lambda, we probably want to drop it beforehand to let the proposer fixed the lambda, and avoiding wasting time till the voting period ended. 

Being able to drop only the ended proposals could be limiting. 

(Tho, this should be rarely since, accidentality write a fail lambda should not happen regularly)

gromakovsky replied:
> I'm thinking that in some scenario where we know that the proposal is going to be accepted, but containing fail lambda, we probably want to drop it beforehand to let the proposer fixed the lambda, and avoiding wasting time till the voting period ended.

Below there is a question about unfreezing tokens in case of failing lambda, I think it's relevant here. There I assumed that we drop only accepted proposals and thus it makes sense to unfreeze all frozen tokens. However, if we can drop potentially negative proposals as well, we may want to reconsider unfreezing.

I think if it's known in advance that the proposal will definitely fail, it simply won't get enough positive votes to pass the quorum threshold and won't be accepted. At least with good UI it sounds like a reasonable assumption.

gromakovsky replied:
> probably this does not increase safety

I think requiring the admin to explicitly specify what proposal they are removing is a good idea for safety. Or maybe show a list of proposals and "DROP IT" button near the oldest one. Without a good UI it can be easy to make a mistake, especially if there are 2 proposals in the same block. Another case when it can increase safety is when there are multiple people/computers with admin access who can potentially call `dropProposals` at the same time.

> now when we represent the proposals queue as Map, checking whether the given element matches the head one seems expensive

I didn't review code in this PR closely, but if you think it's expensive, we can assume that the administrator will use a good UI (we can write a CLI client that would take `proposalKey` and check it somehow) and won't be accessed by more than 1 person concurrently.

Martoon-00 replied:
If we will just remove a proposal by key, without checking it is at the head, this won't be less safe, or not? At least, concurrency issues do not seem to be possible this way.

Also, generally, being able to remove non-head proposals seems to make sense, since if we `flush` by batches, then some intermediate element may turn out to be failing, and we may want to remove it immediately. Though with the current definitions detecting which exactly element is failing may be difficult (the client may need to do `for [n .. 1] dryRunFlush` to find the first failing proposal).

gromakovsky replied:
> the client may need to do for [n .. 1] dryRunFlush to find the first failing proposal

Humanity has invented binary search :trollface:

Martoon-00 replied:
I assumed a smart `for`, even more difficult from implementation point of view :smile_cat:

rinn7e commented:
@Martoon-00 I've made a bunch of changes with change a lot of previous logic, hope to get your review again.

1 question relating to dropping the proposals. How should we deal with the frozen token?
Do we treat this as failing proposals, by simply slash the proposer tokens, and return all voter tokens back?

Martoon-00 commented:
> @Martoon-00 I've made a bunch of changes with change a lot of previous logic, hope to get your review again.
> 
> 1 question relating to dropping the proposals. How should we deal with the frozen token?
> Do we treat this as failing proposals, by simply slash the proposer tokens, and return all voter tokens back?

That's a nice observation. I would also ask @gromakovsky about this, but I suspect that yes. If there is a failing proposal, that's a contract owner to blame, and if he says "there was a mistake and now your funds are lost", no one will be happy ))  Better to return everything to the initial state (though that's a bit sad if we have to write a logic for reversing changes, gonna be a nice pot for bugs)

gromakovsky commented:
I see the following options:
1. Return all tokens that were frozen because the proposal was approved (and failed).
2. Return slashed amount of tokens.
3. Let the administrator specify how much to return as part of the `dropProposals` call.

We can combine them by applying one option to the proposer and a different one to the voters. I will discuss it, but probably I would start with doing (1) for both types of entities (proposer, voters). If a proposal got more positive votes than negative, it's a good proposal even if there is a conflicting one that also got enough votes. So we shouldn't punish anyone by returning less tokens.

rinn7e commented:
> I see the following options:
> 
>     1. Return all tokens that were frozen because the proposal was approved (and failed).
> 
>     2. Return slashed amount of tokens.
> 
>     3. Let the administrator specify how much to return as part of the `dropProposals` call.
> 
> 
> We can combine them by applying one option to the proposer and a different one to the voters. I will discuss it, but probably I would start with doing (1) for both types of entities (proposer, voters). If a proposal got more positive votes than negative, it's a good proposal even if there is a conflicting one that also got enough votes. So we shouldn't punish anyone by returning less tokens.

I'm going with the option 1 for now.

rinn7e commented:
@gromakovsky @Martoon-00 Continue the discussion from this thread: https://github.com/tqtezos/baseDAO/pull/71#discussion_r546663541

I've updated:
- explicitly specify a proposal key for dropProposal
- return all frozen tokens back to voters and proposer
- only affect a proposal that is ended and accepted but not yet flush (or can't be flush)

The only problem now is that the contract size is too large again. 😕

This is due to having more code in `dropProposal` related to checking whether or not it is an accepted proposal. Hope to get some advice on how to optimize this.

rinn7e commented:
> @gromakovsky @Martoon-00 Continue the discussion from this thread: [#71 (comment)](https://github.com/tqtezos/baseDAO/pull/71#discussion_r546663541)
> 
> I've updated:
> 
>     * explicitly specify a proposal key for dropProposal
> 
>     * return all frozen tokens back to voters and proposer
> 
>     * only affect a proposal that is ended and accepted but not yet flush (or can't be flush)
> 
> 
> The only problem now is that the contract size is too large again. confused
> 
> This is due to having more code in `dropProposal` related to checking whether or not it is an accepted proposal. Hope to get some advice on how to optimize this.

I've fixed the CI issues now. This PR is ready for re-review.

Martoon-00 reviewed (commented):
Overall looks good to me, I have just some minor suggestions.

gromakovsky commented on src/Lorentz/Contracts/BaseDAO/Doc.hs:
```diff
@@ -147,6 +148,15 @@ flushDoc = [md|
+dropProposalDoc :: Markdown
+dropProposalDoc = [md|
+  Delete a finished and accepted proposal that is not flushed. The amount of frozen tokens is
+  are returned to the proposer and votes in full. The decision lamda is skip.
```
> votes

voters

gromakovsky commented on src/Lorentz/Contracts/BaseDAO/Types.hs:
```diff
@@ -281,12 +286,13 @@ data Parameter proposalMetadata
   | Set_voting_period VotingPeriod
   | Set_quorum_threshold QuorumThreshold
-  | Flush ()
+  | Flush (Maybe Natural)
```
If we hit the operation size limit and have find time dealing with it, we can remove `Maybe` here because the user usually can just pass a large value instead of `Nothing`. Like, if you pass 1000, most likely all proposals will be processed (even if there are more than 1000 in the contract, which is unlikely, processing more than 1000 will likely fail due to gas costs). Anyway, I don't mind having `Maybe` here for now.

gromakovsky commented on src/Lorentz/Contracts/BaseDAO/Types.hs:
```diff
@@ -211,7 +212,7 @@ data Storage (contractExtra :: Kind.Type) (proposalMetadata :: Kind.Type) = Stor
   , sExtra                     :: contractExtra
   , sProposals                 :: BigMap (ProposalKey proposalMetadata) (Proposal proposalMetadata)
-  , sProposalKeyListSortByDate :: [ProposalKey proposalMetadata] -- Newest first
+  , sProposalKeyListSortByDate :: Map (Timestamp, ProposalKey proposalMetadata) () -- Oldest first
```
It looks isomorphic to a set, why are we using a map here?

rinn7e replied:
Yes, it is isomorphic to a set. This is due to, initially the map was `Map Timestamp ( ProposalKey proposalMetadata)`, but due to key collision problem, it becomes `Map (Timestamp, ProposalKey proposalMetadata) ()`.

Do you think we should change to `Set` or not.

gromakovsky replied:
If we can change it to `Set`, I think we should do it. It should be more efficient I guess (and we are already close to hitting the operation size limit). Also it adds clarity.

rinn7e replied:
I've updated this now.

gromakovsky reviewed (changes requested):
Please also update the specification document. In particular, we should explicitly state in which order proposals are processed (and that proposals from the same block will be ordered by their keys).

rinn7e commented on src/Lorentz/Contracts/BaseDAO/Proposal.hs:
```diff
@@ -597,60 +564,138 @@ unfreezeVoterToken = do
+  unpair
+  swap; dup
+  stackType @(Maybe Natural : Maybe Natural : "currentCount" :! Natural : s)
+  ifSome (do
```
I've updated this, as well as other changes.

gromakovsky commented on docs/specification.md:
```diff
@@ -769,6 +773,25 @@ unit
     - The return amount for each voters is equal to or less than the voter's frozen tokens.
 - The lost of frozen tokens is due to the fact that the administrator has the right to `burn` or `transfer` frozen tokens of any proposers or voters.
 - If proposal is accepted, decision lambda is called.
+- Fails with `NOT_ADMIN` if the sender is not the administrator.
```
Oh, that's somewhat unexpected to me. Why don't we let anyone call `flush`? Is there any harm in letting anyone call it? Ideally it should be called automatically when voting period ends, but since it's not possible, we have to explicitly call it. It doesn't seem to use `SENDER` value for anything. And if a non-admin user calls it before the admin, it will be even better.

It's unrelated to this issue, but if you agree with me, I will create an issue about removing this check or you can do it right in this PR because the change should be small I guess.

rinn7e replied:
I'm good with removing the admin restriction.

rinn7e commented on docs/specification.md:
```diff
@@ -750,15 +751,18 @@ Parameter (in Michelson)
+- Finish voting process on an amount of proposals:
+  - If the parameter is `None`, process on all the proposals for which voting period is over.
+  - If the paramter is `Just n`, process `n` proposals for which voting period is over.
+- The order of processing proposals are from 'the oldest' to 'the newest'.
```
Updated.

gromakovsky reviewed (approved):
My review was not very thorough, but what I've seen looks ok. Hopefully @martoon has reviewed everything I skipped :)
